crates/apollo-router-core/src/query_planner/model.rs

```diff
@@ -70,6 +75,185 @@ impl PlanNode {
             Self::Flatten(flatten) => Box::new(flatten.node.service_usage()),
         }
     }
+
+    /// Recursively validate a query plan node making sure that all services are known before we go
+    /// for execution.
+    ///
+    /// This simplifies processing later as we can always guarantee that services are configured for
+    /// the plan.
+    ///
+    /// # Arguments
+    ///
+    ///  *   `plan`: The root query plan node to validate.
+    pub fn validate_services_against_plan(
+        &self,
+        service_registry: Arc<dyn ServiceRegistry>,
+    ) -> Vec<FetchError> {
+        self.service_usage()
+            .collect::<HashSet<_>>()
```

Comment:
not sure we need to collect here, since we're going through the entire list anyway, we could check in the service registry for each of them, it's just a hashmap lookup

Reply:
I did not change this in this PR. I plan to refactor the `ServiceRegistry` in a future PR to avoid making this one heavier.

Reply:
this can be collected in a `HashSet` after the filter. In the most common case it will be empty, instead of allocating for all services every time

Reply:
Right!! Thanks! 80286fc